modules/FvwmGtk/FvwmGtk.1
.de EX		\"Begin example
.ne 5
.if n .sp 1
.if t .sp .5
.nf
.in +.5i
..
.de EE
.fi
.in -.5i
.if n .sp 1
.if t .sp .5
..
.TH FvwmGtk 1 "7 May 1999"
.UC
.SH NAME
\fBFvwmGtk\fP \- the Fvwm GTK module
.SH SYNOPSIS
\fBFvwmGtk\fP is spawned by fvwm, so no command line invocation will work.
From within the	fvwm's configuration file, \fBFvwmGtk\fP is spawned as follows:
.EX
Module FvwmGtk [ name ]
.EE
or from within an fvwm pop-up menu:
.EX
DestroyMenu Module-Popup
AddToMenu   Module-Popup "Modules" Title
AddToMenu   Module-Popup "Gtk" Module FvwmGtk [ name ]
.EE

.SH DESCRIPTION
The \fBFvwmGtk\fP module implements GTK-based alternatives to the 
GUI elements in fvwm, namely the builtin menus and the FvwmForm dialogs.

.SH INVOCATION
No command line invocation is possible. 
\fBFvwmGtk\fP must be invoked by the
\fBfvwm2\fP window manager.
When invoked with the optional \fIname\fP argument, \fIname\fP
is used to find configuration commands and configuraton files
instead of "FvwmGtk".

.SH CONFIGURATION OPTIONS
\fBFvwmGtk\fP only has options to define the content of the menus and dialogs.
This includes the labels and pixmaps to show; configuration of the appearance 
must be done through the GTK rc file mechanism. 

\fBFvwmGtk\fP reads the same configuration file as \fBfvwm\fP when it starts
up. In addition, \fBFvwmGtk\fP accepts commands from fvwm and its modules as
it runs. 

If the optional \fIname\fP is used to start FvwmGtk, \fIname\fP is used 
in all commands, messages, menus and forms generated by FvwmGtk and in the
configuration file name.  Unlike other fvwm modules, there is little reason
to use the optional name. 

All dialogs and menus have to be defined through configuration commands
before they can be used. A dialog or menu is invoked by sending its name
to FvwmGtk. For menus, you also send the button which is to be used.
.EX
	SendToModule FvwmGtk menu-example 1
	SendToModule FvwmGtk dialog-example
.EE

.SH MENUS
The following commands define menus. For all arguments named "label"
in the following menu commands, FvwmGtk looks for embedded ampersands
in the same way fvwm does for its menus.

.IP "*FvwmGtkMenu \fBname\fP"
Instructs \fBFvwmGtk\fP to append subsequent items to the menu named
by the argument. Note that you can "reopen" a menu and continue to append
to it. It is also possible to redefine a menu from scratch, see 
"*FvwmGtkDestroy".

.IP "*FvwmGtkTitle \fBlabel\fP [ \fBicon\fP [ \fBr_label\fP ] ]"
Appends a title to the currently open menu. If the optional argument
is given, it should be the name of an xpm file in the ImagePath. The 
icon will appear to the left of the text. If FvwmGtk has been compiled
with imlib support, icon can of any image format imlib can read. The label 
can contain an ampersand to mark the following character as an accelerator to
be underlined. Underlined accelerators work in the same way as they do in fvwm
native menus. If r_label is given, it will be right-justified.

.IP "*FvwmGtkItem \fBlabel\fP \fBaction\fP [ \fBicon\fP [ \fBr_label\fP ] ]"
Appends an item to the currently open menu. The first argument is the
text that will appear in the item, the second argument will be sent to
fvwm when the item is activated. If the optional argument is given, it 
should be the name of an xpm file in the ImagePath. The icon will appear 
to the left of the text. The label can contain an ampersand
to mark the following character as an accelerator to be underlined. 
Underlined accelerators work in the same way as they do in fvwm native menus.
If r_label is given, it will be right-justified.

.IP "*FvwmGtkSubmenu \fBlabel\fP \fBname\fP [ \fBicon\fP ]"
Appends an item to the currently open menu. The first argument is the
text that will appear in the item, the second argument is the name of 
the submenu that will be opened when the item is selected. If the submenu 
doesn't exist yet, it will be created. If the optional argument is given, 
it should be the name of an xpm file in the ImagePath. The icon will 
appear to the left of the text. The label can contain an ampersand
to mark the following character as an accelerator to be underlined. 
Underlined accelerators work in the same way as they do in fvwm native menus.

.IP "*FvwmGtkTearoff"
Appends a special tear-off item. Activating it turns the menu into 
a permanent window; activating it once more makes the menu disappear again. 

.SH WINDOW LISTS

Window lists are special dynamically created menus. They offer a list of 
all windows managed by fvwm and send "WindowListFunc window-id" back to 
fvwm, where window-id is the ID of the selected window in hexadecimal 
notation as expected by fvwm. WindowListFunc is the same function that is 
also used by fvwm for its builtin window list.

.IP "*FvwmGtkWindowList \fBname\fP [ \fBoption\fP... ]"
Creates a window-list bound to \fBname\fP. The window-list will be 
formatted according to the given options. Currently supported are
the following options which have the same meaning as for the fvwm's 
builtin window list:
"NoGeometry", "NoMiniIcon", "UseIconName", "Desk <desk>", "CurrentDesk", 
"Icons/NoIcons/OnlyIcons", "Sticky/NoSticky/OnlySticky", 
"Normal/NoNormal/OnlyNormal", "NoDeskSort", "Alphabetic", 
"Function <func>".
The following options are new: "Title <label> <icon> <r_label>".
Additionally, you can specify a name pattern to match the name or
icon name of the listed windows.


.SH DIALOGS
Dialogs consist of hierarchically organized widgets. Widgets are either 
composite, ie they can contain one or more other widgets, or atomic. 
Composite widgets are defined by a pair of commands denoting the start 
and end of the list of child widgets. Atomic widgets are defined by 
single commands. Some widgets request data from the user. These are 
all given names which can be used to refer to the data. 

The commands for widgets have the common form

"*FvwmGtk\fIWidget\fP \fIwidget-specific args\fP [ -- \fIgeneral args\fP ]"

where the following arguments may be specified after the double hyphen:
can-default, default, focus, expand, fill and an integer argument padding.

\fIcan-default\fP means that the widget may become the default widget
of the dialog (ie the one being activated if the user immediately presses
Return).

\fIdefault\fP means that the widget is the initial default widget.

\fIfocus\fP means that the widget initially has the keyboard focus.

\fIexpand\fP, \fIfill\fP, \fIpadding\fP are relevant if the widget
is part of a row or column (see the GTK documentation). 

The following paragraphs explain only the widget-specific arguments.

.IP "*FvwmGtkDialog \fBname\fP \fBtitle\fP [ \fBcenter\fP ]"
Starts or reopens a dialog named by the first argument. \fBtitle\fP
is used as the window title. If the optional argument \fBcenter\fP is given,
the dialog will be mapped at the center of the screen. Otherwise it
will be mapped at the mouse position.

A dialog has one child widget.

.IP "*FvwmGtkBox [ \fBvertical\fP ] [ \fBhomogeneous\fP ] [ \fBspacing\fP [ \fBborder\fP ] ]"
.IP "*FvwmGtkEndBox"
Start and end a composite widget which is used to group a number of child
widgets in a row or column. The arguments influence the placement of the child
widgets (for information about their meaning, see the GTK documentation).

.IP "*FvwmGtkFrame \fBlabel\fP [ \fBborder\fP ]"
.IP "*FvwmGtkEndFrame"
Start and end a composite widget which draws a frame labelled by the given
label around a single child widget. The optional argument determines
the separation between the frame and the child.

.IP "*FvwmGtkLabel \fBlabel\bP"
Adds a label widget.

.IP "*FvwmGtkEntry \fBname\fP [ \fBinitial-value\fP ]"
Adds a widget which allows the user to enter a string. If given, the 
\fBinitial-value\fP is displayed initially. The acutal string can
be referred to by \fBname\fP. 

.IP "*FvwmGtkButton \fBlabel\fP \fBcmd\fP ..."
Adds a button widget with the given label which will trigger the given
commands when pressed. Commands can contain references to the values
of widgets in the form $(name). These references are replaced by the actual
values. The replacement is done recursively, ie the values can again
contain references.

Commands can be strings to send to fvwm, system commands starting with
and exclamation sign (which is stripped before execution) or the special
command "close" to close the current dialog.

.IP "*FvwmGtkCheckButton \fBname\fP \fBlabel\fP \fBon-value\fP \fBoff-value\fP [ \fBon\fP ]"
Adds a check button widget with the given label. The value referred to
by \fBname\fP will be either \fBon-value\fP or \fBoff-value\fP, depending
on the state of the check button. The initial state is off, unless
the optional argument is given.

.IP "*FvwmGtkRadioGroup \fBname\fP"
.IP "*FvwmGtkEndRadioGroup"
Start and end a composite widget which groups a number of mutually 
exclusive radio buttons. The value of the enables radio button can be 
referred to by \fBname\fP. Grouping radio buttons in this way doesn't 
influence their geometry. You can put a radio group in a row or column 
or even distribute it across several rows to determine the geometry.

.IP "*FvwmGtkRadioButton \fBlabel\fP \fBon-value\fP [ \fBon\fP ]"
Adds a radio button with the given label. Its initial state is off, unless
the optional argument is given. There should probably be exactly one \fBon\fP
radio button in a radio group. A radio button must be child of a radio group
in order to be able to access its value.  

.IP "*FvwmGtkNotebook \fBlabel\fP"
Opens a new notebook page with the given label. If there is already an open
notebook, the page is appended to it. Otherwise a new notebook is created. A
notebook page is a composite widget that expects one child.

.IP "*FvwmGtkEndNotebook"
Closes a notebook widget. 

.IP "*FvwmGtkColor \fBname\fP [ \fBinitial-value\fP ]"
Adds a color selector whose initial color is specified by \fBinitial-value\fP
and whose value can be referred to by \fBname\fP. The \fBinitial-value\fP
can be any color specification that is accepted by XParseColor. The returned
color specifications are always of the form "rgb:rrrr/gggg/bbbb".

.IP "*FvwmGtkScale \fBname\fP [ \fBvertical\fP ] \fBvalue\fP \fBlower\fP
\fBupper\fP \fBinc\fP \fBpage-inc\fP \fBpage-size\fP [ \fBdigits\fP ]"
Adds a scale. 

.IP "*FvwmGtkSpinButton \fBname\fP \fBvalue\fP \fBlower\fP
\fBupper\fP \fBinc\fP \fBpage-inc\fP \fBpage-size\fP \fBclimb-rate\fP [
\fBdigits\fP ]"
Adds a spin button.

.IP "*FvwmGtkOptionMenu \fBname\fP"
.IP "*FvwmGtkEndOptionMenu"
.IP "*FvwmGtkItem \fBlabel\fP \fBvalue\fP [ \fBon\fP ]"
Can be used to create option menus.


.SH COMMON CONFIGURATION
The following commands can be used for menus and dialogs.

.IP "*FvwmGtkSeparator"
Appends a separator to the currently open menu, row or column.

.IP "*FvwmGtkDestroy \fBname\fP"
Destroys the specified menu or dialog. 

.IP "*FvwmGtkRCFile \fBfile\fP"
Note that this command should be issued before defining any menus
or dialog. Hint for GNOME users: If you add instances of this command for the
standard GNOME rc files, switching themes via the control-center will
apply to FvwmGtk widets as well, giving a very integrated appearance
of the desktop.

.IP "*FvwmGtkIconSize [ \fBwidth\fP \fBheight\fP ]"
If FvwmGtk has been compiled with imlib support, icons will be scaled
to the size set by this command. Omitting the arguments disables scaling. 
Note that there is currently no way to change the amount of room left 
free for icons, thus using a width of more than 20 will not be very
useful.

.SH COMMANDS
To invoke one of the previously defined menus or dialogs, use the fvwm2 
command \fISendToModule\fP.

.IP "SendToModule FvwmGtk \fBname\fP \fBbutton\fP"
makes \fBFvwmGtk\fP pop up the menu or dialog named by the first argument.
The second argument is the button that will be used for menu selection.
For dialogs and menus which are not popped up from a button press, you can
omit the button.  
Examples:
.EX
Mouse 3 R A SendToModule FvwmGtk Window-Ops 3
Key F10 R A SendToModule FvwmGtk Applications-Menu
Mouse 1 R A SendToModule FvwmGtk Quit-Verify-Dialog
.EE

.SH AUTHOR
\fIMatthias Clasen\fP <clasen@mathematik.uni-freiburg.de>

.SH ACKNOWLEDGMENTS
The gtkpixmapmenuitem code and the code for underline accelerators is 
taken from libgnomeui.

.SH BUGS
\fBFvwmGtk\fP should be turned into a Gnome application.
The builtin menus offer many features which the GTK menus currently
don't have. None of the features of the builtin window-list have
been implemented.



NEWS
Note, the changes for the last STABLE release start with release
2.6.0.

-------------------------------------------------------------------
Changes in stable release 2.6.8 (UNRELEASED)

* Bug fixes:

  - Various DESTDIR fixes (especially around the default-config
    Makefile)
  - fvwm-perllib pod2man fixes.
  - FvwmIconMan no longer triggers a warning about bad size hints
    in the fvwm core.
  - VMware windows do not disapper when switching between
    fullscreen and normal state.
  - Fix (de)installation with the configure options
    --program-prefix, --program-suffix and
    --program-transform-name.

-------------------------------------------------------------------
Changes in stable release 2.6.7 (06-Mar-2016)

* New fvwm features:

  - A new default configuration which is available when fvwm
    doesn't detect a configuration file to load.
  - A new conitional command "Desk n" can restrict matching
    windows to a specific desk.
  - A new conditional command "Screen n" to restrict matching
    windows on a given Xinerama screen.
  - New expansion variable "w.screen" to ascertain the Xineram
    screen number a window is on.
  - New command "InfoStoreClear" to remove all items in the
    InfoStore.

* Removed features:

  - The old and unmaintained debian/and rpm/ directories have
    been remmoved; use the maintainers' copies where available.
  - VMS support has been removed.
  - GTK1.x support has been removed.
  - GNOME-specific window hints (pre-EWMH) have been removed.
  - Some fvwm modules have been removed:
      - FvwmDragWell   (no replacement)
      - FvwmGTK        (no replacement)
      - FvwmSave       (no replacement)
      - FvwmSaveDesk   (no replacement)
      - FvwmScroll     (no replacement)
      - FvwmTabs       (no replacement, never worked anyway)
      - FvwmTaskBar    (use FvwmButtons)
      - FvwmTheme      (in core of fvwm as colorsets)
      - FvwmWharf      (use FvwmButtons)
      - FvwmWinList    (use WindowList command)
      - FvwmWindowMenu (use WindowList command)
      - FvwmIconBox    (use the IconBox style instead)

* New module features:

  - FvwmButtons learned a new option "Colorset" to its
    ChangeButton command.

* fvwm-menu-desktop updated:

  - Renamed default menu to XDGMenu and changed the name
    of the FvwmForm to FvwmForm-XDGMenu-Config to not conflict
    with someone already using FvwmMenu.
  - fvwm-menu-desktop will now load defaults from the
    FvwmForm-XDGMenu-Config data file.
  - Improved dynamic menus by regenerating them on-the-fly.
  - Added new options: --regen-cmd, --dynamic, and more.

* Bug fixes:

  - A bug introduced in 2.6.6 could cause applications with
    negative coordinates to be placed at strange positions.  This
    affected for example acroread when switching to fullscreen
    mode.  This has been fixed.
  - FvwmButtons "Silent" option for dynamic ChangeButton commands
    no longer loops infinitely.

Changes in stable release 2.6.6 (15-Mar-2016)

* New fvwm features:

  - Support for Russian from Ivan Gayevskiy.
  - EnvMatch supports infostore variables.
  - The option "forget" to the Maximize command allows to
    unmaximize a window without changing its size and position.
  - Windows shaded to a corner (NW, NE, SW, SE) are reduced to a
    small square.
  - New option "!raise" to the WarpToWindow command.
  - The new extended variables $[wa.x], $[wa.y], $[wa.width],
    $[wa.height] can be used to get the geometry of the EWMH
    working area, and $[dwa.x], $[dwa.y], $[dwa.width],
    $[dwa.height] can be used to get the geometry of the EWMH
    dynamic working area.
  - The Resize commands accept "wa" or "da" as a suffix of the
    width or height value.  If present, the value is a percentage
    of the width or height of the EWMH working area or the EWMH
    dynamic working area.
  - Fvwm is much more resilient against applications that flood the
    window manager with repeated events.

* New module features:

  - FvwmForm supports separator lines.
  - New FvwmIconMan options:
      IconAndSelectButton
      IconAndSelectColorset

* Bug fixes:

  - Provide a wrapper for the deprecation of XKeycodeToKeysym and
    use XkbKeycodeToKeysym() where appropriate.
  - fvwm-menu-desktop is re-written and provides better support of
    the XDG menu specification.
  - Fix fvwm-menu-desktop keyError bug.  Use "others" if no desktop
    environment found.
  - FvwmIconMan had problems displaying the hilight colour on some
    systems (64 bit issue?).
  - Globally active windows cannot take the focus if the style
    forbids programs to take focus themselves (style
    !FPFocusByProgram).
  - Windows no longer jump from one position to the other which
    could happen in some cases with SnapAttraction.  Windows now
    snap to the closest window (or screen edge).
  - Removing bindings had several strange side effects that are
    fixed now (removing too many bindings; old bindings showing up
    again after another is removed; possibly other effects).
  - Windows sometimes did not get expose events (i.e. did not
    redraw properly) if they were uncovered by moving a window
    above them.  This has been fixed.
  - FvwmConsole now causes much less network traffic.
  - Suppress bogus events sent to the modules when a window is
    resized with the mouse.
  - Properly handle the has_ref_window_moved flag for ResizeMove and
    ResizeMoveMaximize.
  - Removed some unnecessary redraws in FvwmPager.
  - The option "NoDeskLimitY" option of the GotoPage command did
    not work.
  - Negative coordinates in the "rectangle" option to the Menu
    commend did not work correctly.  This has been fixed.
  - Removes a slight graphics problem whith the ResizeMaximize
    command being invoked from a window button menu.
  - When an attempt to reparent a client window (i.e. decorate
    it) fails, fvwm no longer throws away all events but only the
    events for that window.

modules/FvwmGtk/FvwmGtk.c
/* -*-c-*- */
/*
 * FvwmGtk - gtk menus and dialogs for fvwm
 *
 * Copyright (c) 1999 Matthias Clasen <clasen@mathematik.uni-freiburg.de>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#include "config.h"

#include <stdio.h>
#include <unistd.h>

#include <X11/Xlib.h>

#ifdef NEED_GNOMESUPPORT_H
#include <gnome.h>
#else
#include <glib.h>
#include <gtk/gtk.h>
#endif

#ifdef GDK_IMLIB
#include <gdk_imlib.h>
#endif

#include <libs/Module.h>
#include <libs/fvwmlib.h>
#include <fvwm/fvwm.h>
#include <libs/vpacket.h>
#include "libs/Parse.h"
#include "libs/Strings.h"

#include "dialog.h"
#include "menu.h"
#include "windowlist.h"


int fvwm_fd[2] = { -1, -1 };
static ModuleArgs *module;
char *image_path = NULL;
GHashTable *widgets = NULL;
GtkWidget *current = NULL;
GtkWidget *attach_to_toplevel = NULL;
unsigned long context = 0;
int icon_w = 0;
int icon_h = 0;
long current_desk = 0;
GHashTable *window_list_entries = NULL;

/*
   General overview:

   All widgets (menus and dialogs) are stored in one hash table, widgets.
   They are built up piecemeal, the widget currently being built up is current.
   For dialogs, current may also point to a subwidget of the dialog.

   dialog contain diffent kinds of widgets:

   * data widgets which return some value these all have a name and store
   their value as data on the top-level dialog under their name

   * action widgets like buttons. These have a list of commands which are
   sent to fvwm or executed by FvwmGtk itself (eg close).
   The commands are stored as data on the action widget under the name
   "values", as a NULL-terminated char*-array.

   A command can contain references to other widgets values in the form
   $(name). These references are resolved recursively, ie the values can
   again contain references.

   The resolution of the references is done by splitting the string into a
   list of passive strings and variable references and splicing in sublists
   for the variable references until only passive strings remain.
 */


void destroy(int argc, char **argv)
{
	GtkWidget *w;

	g_return_if_fail(argc >= 1);

	w = g_hash_table_lookup(widgets, argv[0]);
	if (w != NULL)
	{
		if (gtk_widget_get_toplevel(current) == w)
		{
			current = NULL;
		}

		g_hash_table_remove(widgets, argv[0]);
		gtk_object_unref(GTK_OBJECT(w));
		gtk_widget_destroy(w);
	}
}


void separator(int argc, char **argv)
{
	if (GTK_IS_MENU(current))
	{
		menu_separator(argc, argv);
	}
	else
	{
		dialog_separator(argc, argv);
	}
}


void item(int argc, char **argv)
{
	if (GTK_IS_MENU(current))
	{
		  menu_item(argc, argv);
	}
	else
	{
		  dialog_option_menu_item(argc, argv);
	}
}


void parse_rc_file(int argc, char **argv)
{
	g_return_if_fail(argc >= 1);

	gtk_rc_parse(argv[0]);
}


void icon_size(int argc, char **argv)
{
	if (argc < 2)
	{
		icon_w = 0;
		icon_h = 0;
	}
	else
	{
		icon_w = atoi(argv[0]);
		icon_h = atoi(argv[1]);
	}
}


char *table[] = {
	"Box",
	"Button",
	"CheckButton",
	"Color",
	"Destroy",
	"Dialog",
	"EndBox",
	"EndFrame",
	"EndNotebook",
	"EndOptionMenu",
	"EndRadioGroup",
	"Entry",
	"Frame",
	"IconSize",
	"Item",
	"Label",
	"Menu",
	"Notebook",
	"OptionMenu",
	"RadioButton",
	"RadioGroup",
	"RCFile",
	"Scale",
	"Separator",
	"SpinButton",
	"Submenu",
	"Tearoff",
	"Title",
	"WindowList",
};


void (*handler[])(int, char**) = {
	dialog_start_box,
	dialog_button,
	dialog_checkbutton,
	dialog_color,
	destroy,
	open_dialog,
	dialog_end_something,
	dialog_end_something,
	dialog_end_notebook,
	dialog_end_option_menu,
	dialog_end_radiogroup,
	dialog_entry,
	dialog_start_frame,
	icon_size,
	item,
	dialog_label,
	open_menu,
	dialog_notebook,
	dialog_start_option_menu,
	dialog_radiobutton,
	dialog_start_radiogroup,
	parse_rc_file,
	dialog_scale,
	separator,
	dialog_spinbutton,
	menu_submenu,
	menu_tearoff_item,
	menu_title,
	window_list
};


void parse_arguments(char **line, int *argc, char ***argv)
{
	char *tmp[100];
	int i;

	for (i = 0; i < 100 ; i++)
	{
		*line = GetNextSimpleOption(*line, &tmp[i]);
		if (!tmp[i]) break;
	}
	*argc = i;

	*argv = (char **) safemalloc(i * sizeof(char *));
	for (i = 0; i < *argc; i++)
	{
		(*argv)[i] = tmp[i];
	}
}


void widget_not_found(char *name)
{
	GtkWidget *dialog, *box, *item;
	char buf[200];

	SendText(fvwm_fd, "Beep", 0);
	dialog = gtk_window_new(GTK_WINDOW_DIALOG);
	gtk_window_set_title(GTK_WINDOW(dialog), "Error");
	gtk_window_set_wmclass(GTK_WINDOW(dialog), "Error", "FvwmGtk");

	box = gtk_vbox_new(FALSE, 10);
	gtk_container_add(GTK_CONTAINER(dialog), box);
	gtk_container_border_width(GTK_CONTAINER(dialog), 30);
	g_snprintf(buf, sizeof(buf), "No such menu or dialog: %s", name);
	item = gtk_label_new(buf);
	gtk_box_pack_start(GTK_BOX(box), item, FALSE, FALSE, 5);
	item = gtk_button_new_with_label("OK");
	gtk_box_pack_start(GTK_BOX(box), item, FALSE, FALSE, 5);
	gtk_signal_connect_object(
		GTK_OBJECT(item), "clicked",
		GTK_SIGNAL_FUNC(gtk_widget_destroy), GTK_OBJECT(dialog));
	gtk_widget_show_all(box);
	gtk_widget_show(dialog);
}


void parse_config_line(char *buf)
{
	int argc;
	char **argv;
	char *p;
	char **e;

	if (buf[strlen(buf)-1] == '\n')
	{
		buf[strlen(buf)-1] = '\0';
	}
	if (
		strncasecmp(
			buf, CatString3("*",module->name,0),
			module->namelen+1) == 0)
	{
		p = buf + module->namelen+1;
		if ((e = FindToken(p, table, char*)))
		{
			p += strlen(*e);
			parse_arguments(&p, &argc, &argv);
			handler[e - (char**)table](argc, argv);
		}
		else
		{
			fprintf(stderr, "%s: unknown command: %s\n",
				module->name, buf);
		}
	}
	else if (strncasecmp(buf, "ImagePath", 9) == 0)
	{
		if (image_path)
		{
			free(image_path);
		}
		image_path = stripcpy(buf + 9);
	}
}


void parse_options(void)
{
	char *buf;

	/* only my config lines needed */
	InitGetConfigLine(fvwm_fd,CatString3("*",module->name,0));
	while (GetConfigLine(fvwm_fd, &buf), buf != NULL)
	{
		parse_config_line(buf);
	}
}


void process_message(
	unsigned long type, unsigned long timestamp, unsigned long *body)
{
	int button = 0;
	GtkWidget *widget;
	window_list_options *opts;

	char name[128];

	switch (type)
	{
	case M_STRING:
		SendUnlockNotification(fvwm_fd);
		context = body[0]; /* this is fw */
		sscanf((char*)(&body[3]), "%127s %d", name, &button);
		widget = g_hash_table_lookup(widgets, name);
		if (!widget)
		{
			widget_not_found(name);
		}
		else if (GTK_IS_MENU(widget))
		{
			opts = (window_list_options *)gtk_object_get_data(
				GTK_OBJECT(widget), "window_list");
			if (opts)
			{
				char *argv[1];

				argv[0] = name;
				destroy(1, argv);
				open_menu(1, argv);
				widget = current;
				gtk_object_set_data(
					GTK_OBJECT(current), "window_list",
					opts);
				construct_window_list();
			}

			gtk_menu_popup(
				GTK_MENU(widget), NULL, NULL, NULL, NULL,
				button, timestamp);
		}
		else if (GTK_IS_WINDOW(widget))
		{
			gtk_widget_show(GTK_WIDGET(widget));
		}
		break;
	case M_CONFIG_INFO:
		parse_config_line((char *)(&body[3]));
		break;
	case M_NEW_DESK:
		current_desk = (long)body[0];
		break;
	case M_ADD_WINDOW:
	case M_CONFIGURE_WINDOW:
		{
			struct ConfigWinPacket *cfg = (void *)body;
			window_list_entry *wle =
				lookup_window_list_entry(body[0]);

			wle->desk = cfg->desk;
			wle->layer = 0;
			wle->iconified = IS_ICONIFIED(cfg);
			wle->sticky = (IS_STICKY_ACROSS_PAGES(cfg) ||
				       IS_STICKY_ACROSS_DESKS(cfg));
			wle->skip = DO_SKIP_WINDOW_LIST(cfg);
			wle->x = cfg->frame_x;
			wle->y = cfg->frame_y;
			wle->width = cfg->frame_width;
			wle->height = cfg->frame_height;
		}
		break;
	case M_DESTROY_WINDOW:
		g_hash_table_remove(window_list_entries, &(body[0]));
		break;
	case M_VISIBLE_NAME:
		{
			window_list_entry *wle =
				lookup_window_list_entry(body[0]);
			if (wle->name)
			{
				free(wle->name);
			}
			wle->name = safestrdup((char *)(&body[3]));
		}
		break;
	case MX_VISIBLE_ICON_NAME:
		{
			window_list_entry *wle =
				lookup_window_list_entry(body[0]);
			if (wle->icon_name)
			{
				free(wle->icon_name);
			}
			wle->icon_name = safestrdup((char*)(&body[3]));
		}
		break;
	case M_MINI_ICON:
		{

			MiniIconPacket *mip = (MiniIconPacket *)body;
			window_list_entry *wle =
				lookup_window_list_entry(mip->w);

			if (wle->mini_icon)
			{
				free(wle->mini_icon);
			}
			wle->mini_icon = mip->name[0] != '\0' ?
				safestrdup(mip->name) : NULL;
		}
		break;
	}
}


void read_fvwm_pipe(gpointer data, int source, GdkInputCondition cond)
{
	FvwmPacket* packet = ReadFvwmPacket(source);
	if (packet == NULL)
	{
		exit(0);
	}
	else
	{
		process_message(packet->type, packet->timestamp, packet->body);
	}
}

int main(int argc, char **argv)
{
	module = ParseModuleArgs(argc,argv,1);
	if (module == NULL)
	{
		fprintf(
			stderr, "FvwmGtk version %s should only be executed by"
			" fvwm!\n", VERSION);
		exit(1);
	}

	fvwm_fd[0] = module->to_fvwm;
	fvwm_fd[1] = module->from_fvwm;

#ifdef GDK_IMLIB
	gdk_init(&argc, &argv);
	gdk_imlib_init();
	gtk_widget_push_visual(gdk_imlib_get_visual());
	gtk_widget_push_colormap(gdk_imlib_get_colormap());
#endif

#ifdef NEED_GNOMESUPPORT_H
	gnome_init("FvwmGtk", VERSION, argc, argv);
	gnome_client_disconnect(gnome_master_client());
#else
	gtk_init(&argc, &argv);
#endif

	widgets = g_hash_table_new(g_str_hash, g_str_equal);

	attach_to_toplevel = gtk_window_new(GTK_WINDOW_TOPLEVEL);
	gtk_widget_realize(attach_to_toplevel);

	window_list_entries = g_hash_table_new(g_int_hash, g_int_equal);

	parse_options();

	/* normal messages */
	SetMessageMask(
		fvwm_fd,
		M_STRING |
		M_CONFIG_INFO |
		M_SENDCONFIG |
		M_ADD_WINDOW |
		M_DESTROY_WINDOW |
		M_CONFIGURE_WINDOW |
		M_NEW_DESK |
		M_VISIBLE_NAME |
		M_MINI_ICON);
	/* extended messages */
	SetMessageMask(
		fvwm_fd,
		MX_VISIBLE_ICON_NAME);

	SendText(fvwm_fd, "Send_WindowList", 0);

	gtk_input_add_full(
		fvwm_fd[1], GDK_INPUT_READ | GDK_INPUT_EXCEPTION,
		read_fvwm_pipe, NULL, NULL, NULL);

	/* tell fvwm we're running */
	SendFinishedStartupNotification(fvwm_fd);

	/* tell fvwm we want to be lock on send for M_STRING Messages */
	SetSyncMask(fvwm_fd, M_STRING);

	gtk_main();
	return 0;
}

